Transpose rule for 'allreduce_mpi' (with MPI.SUM)

inailuig opened the pull request:
Adresses https://github.com/PhilipVinc/mpi4jax/issues/34.

inailuig commented on mpi4jax/collective_ops/allreduce.py:
```diff
@@ -52,7 +52,7 @@ def Allreduce(x, op, comm=_MPI.COMM_WORLD, token=None):
     if token is None:
-        token = create_token(stop_gradient(x))
+        token = create_token(x)
 
```
`create_token` already calls `stop_gradient` internally, see https://github.com/google/jax/blob/8f0425d4120a8a3858914c1f4c73c50d4b484ae6/jax/_src/lax/lax.py#L5858
so I think we can omit it here?

PhilipVinc replied:
I think that's new. 
Yes, we can drop it.

PhilipVinc commented:
What you did is quite nice, and maybe we can also support nested transposition so reason just a moment with me...

Normally AllReduce is a 2 input -> 2 output function,  the token and the actual input. Likewise, the transpose rule will take as input 2 duals (Cotangents) and return two duals (tangents), according to the identity.

So re-transposingg this is say and no prob.

However, sometimes Allreduce will take only 1 input and return two outputs, because it creates internally the token.
So this is equivalent to 

`Allreduce_no_token(x) = Allreduce(x, create_token(x)) `

So we only need to really write a transpose rule for `create_token` and everything will be ok.
Now, create token takes an input and gives a token, creating a (fake) dependency in the DAG between the input `x` and the token.

A (fake) yet valid transposition would be to return as a transpose a hard Zero().
Do you agreeE?

PhilipVinc commented:
taking your example,
why don't we return instead of None

```python
def _stop_gradient_transpose_rule(primals, tangents):
    return (primals,)

def _create_token_transpose_rule(primals, tangents):
    return (None,)

from jax import ad_util
from jax import lax
ad.primitive_transposes[ad_util.stop_gradient_p] = _stop_gradient_transpose_rule
ad.primitive_transposes[lax.create_token_p] = _create_token_transpose_rule
```

Why don't we return instead of None a hard zero (`ad.zero`)?

dionhaefner commented:
IMO instead of introducing all of these workarounds and hacks, I would prefer to require people to pass in the token, so we don't need to mess with the create_token primitive.

dionhaefner commented:
Could also be worth to escalate this to the JAX people, there might be a better way to define trivial transposes.

inailuig commented:
I also added the jvp rules for allreduceT (and corrected the one for allreduce; it was the vjp rule and not jvp). Thanks to the transpose rules we now have both jvp and vjp.

Can someone check my usage of the tokens? The way I am using it is hiding the passing of the token in a closure, and always reusing that in initial token; see https://github.com/inailuig/mpi4jax/commit/b782b2e89c97bff3e768ffc1654227075baae71c

PhilipVinc commented on mpi4jax/collective_ops/allreduce.py:
```diff
@@ -161,24 +161,115 @@ def mpi_allreduce_value_and_jvp(in_args, tan_args, op, comm):
-    if op == _MPI.SUM:
-        jvp = (x_tan, token_tan)
+    if op_ == _MPI.SUM:
+        jvp = mpi_allreduce_p.bind(x_tan, token, op=op, comm=comm)
```
Are you sure about this? 
Because it seems to me in `res = Allreduce(x, token=token, op=op, comm=comm)` you are essentially calling `mpi_allreduce_p.bind(x, token, op comm)` while there you are calling `mpi_allreduce_p.bind(x_tan, token, op=op, comm=comm)`. Shouldn't you be binding to the transpose operation? Or it's smart enough that if the type of inputs are cotangents, he knows it has to evaluate the transpose?

inailuig replied:
I think its correct like this. For linear operators (which the sum is) the jacobian is just the matrix representing the operator, so the jvp is the same as the operator (and the vjp is the same as the transposed operator). I think what is happening when you call the vjp is that jax calculates the transpose of the jvp (which is linear by definition) using our rule.

inailuig replied:
I just was not sure if we should use token_tan instead of token here?

PhilipVinc commented on mpi4jax/__init__.py:
```diff
@@ -23,6 +23,7 @@
 
 __all__ = [
     "Allreduce",
+    "AllreduceT",
```
surely we don't need to export such implementation detail. also above

inailuig replied:
I think we need to export it, since we want our users to be able to also define the transposed version of matrix-vector multiplication which transposes correctly, see `matvec_transpose_mpi` in https://github.com/PhilipVinc/mpi4jax/issues/34#issuecomment-742454368

(in netket we would need it here: https://github.com/netket/netket/blob/fabb3c55fa45387cbe2080359e4cf6901ac80e45/netket/optimizer/jax/stochastic_reconfiguration.py#L21)

PhilipVinc replied:
ok, you convinced me.

PhilipVinc commented on mpi4jax/collective_ops/allreduce.py:
```diff
@@ -161,24 +161,115 @@ def mpi_allreduce_value_and_jvp(in_args, tan_args, op, comm):
+    comm=(_MPI.Intracomm, HashableMPIType),
+    token=(type(None), xla.Token, core.Tracer),
+)
+def AllreduceT(x, op, comm=_MPI.COMM_WORLD, token=None):
+    if token is None:
+        token = create_token(x)
```
A user will never call this, so this is only called internally.
as, internally, there is always a token, we don't need the default `token=None`, and will never be handling that case, right?

PhilipVinc commented on mpi4jax/collective_ops/allreduce.py:
```diff
@@ -161,24 +161,115 @@ def mpi_allreduce_value_and_jvp(in_args, tan_args, op, comm):
+mpi_allreduceT_impl = default_primitive_impl(mpi_allreduceT_p)
+
+
+def mpi_allreduceT_abstract_eval(xs, token, op, comm):
+    return (
+        abstract_arrays.ShapedArray(xs.shape, xs.dtype),
+        abstract_arrays.abstract_token,
+    )
```
Why redefine it? can't you just reuse the same rule defined before?

PhilipVinc commented:
@inailuig can you check my commit [here](https://github.com/PhilipVinc/mpi4jax/commit/30d39a672eef7fed80677707d8e0692b5fc0cb94) ? I believe this would be a much cleaner implementation... but can you check that it works?

PhilipVinc commented:
@dionhaefner Agree. I'll find the time to discuss with them.

inailuig commented:
> @inailuig can you check my commit [here](https://github.com/PhilipVinc/mpi4jax/commit/30d39a672eef7fed80677707d8e0692b5fc0cb94) ? I believe this would be a much cleaner implementation... but can you check that it works?
> 
Works, and reduces the code duplication quite a bit.

inailuig commented on mpi4jax/collective_ops/allreduce.py:
```diff
@@ -133,51 +150,77 @@ def mpi_allreduce_xla_encode_gpu(c, x, token, op, comm):
+            )
+
+        raise NotImplementedError("TODO implement on GPU transpose of Allreduce")
+    else:
```
Wouldn't the same from the cpu version also work here?
If so a disadvantage of having just one primitive is that we have to duplicate this part for every platform.
In https://github.com/PhilipVinc/mpi4jax/pull/35/commits/5a60eaaf8f4c172a9d0d7db4e2be4ef75654be20 I had added it to `xla.translations` directly, because I was assuming it would work.

PhilipVinc replied:
Yes, probably. I simply did not try.

PhilipVinc commented:
@inailuig Great job, thanks!

PhilipVinc commented on mpi4jax/collective_ops/allreduce.py:
```diff
@@ -133,51 +150,77 @@ def mpi_allreduce_xla_encode_gpu(c, x, token, op, comm):
+                "The linear transpose of Allreduce for {} is not defined".format(op)
+            )
+
+        raise NotImplementedError("TODO implement on GPU transpose of Allreduce")
```
```suggestion
        return _ops.Tuple(c, [x, token])
```

inailuig commented on mpi4jax/collective_ops/allreduce.py:
```diff
@@ -51,16 +51,25 @@ def Allreduce(x, op, comm=_MPI.COMM_WORLD, token=None):
+    # keep track of whever we are computing the forward or transposition of
+    # Allreduce, because we are 'cheating' and not performing MPI operations
+    # on the tranposed (even though, in principle, we should) in order
+    # to be more efficient.
```
Which operations are meant by `even though, in principle, we should` ?

PhilipVinc replied:
I mean, in theory you should collect the output of all ranks on every rank (Which is a MPI.Gather operation), and then use all that information to compute the transpose pass.

But as all ranks have the same output, we don't need to do it.

Think for example at implementing the transpose for MPI.Reduce, which gives the same output of Allreduce but only on one rank. In principle you'd need to Send the output  of MPI.Reduce present on the master rank to all other ranks
